R/util_data_summaries.R

```diff
@@ -126,3 +155,509 @@ ds.summarize_by_column <- function(
+    #   up to 1 2-way interaction between iv and moderator
+    #   no other interaction terms can be specified via "*"
+    #       to use interactions in covariates, create manually
+    
+    varlist <- ds.helper$glm1_formula_to_varlist(glm_model$formula)
+    iv <- glm_model$model[,varlist$iv]
+    iv_var_type <- ds.helper$variable_type(iv)
+    is_valid <- FALSE
+    
+    # check iv suitability
+    if( iv_var_type %in% c("boolean","numeric") ){
+        is_valid <- TRUE
+    } else{
+        "ds.glm1 spec only allows ivs and moderators of type " %+% 
+            "boolean or numeric. Supplied iv is: "  %+% iv_var_type %+%
+            " in formula: " %+% glm_model$formula %>%
+            util.warn()
+    }
+    if( ! is.na(varlist$mod) ){ 
+        mod <- glm_model$model[,varlist$mod]
+        mod_var_type <- ds.helper$variable_type(mod)
+        if( ! mod_var_type %in% c("boolean","numeric") ){
+            "ds.extract_glm1_stats only supports mods of type " %+% 
+                "boolean or numeric. Supplied mod is: " %+% mod_var_type %+%
+                " in formula: " %+% glm_model$formula %>%
+                util.warn()
+            is_valid <- FALSE
+        }
+    }
+    return(is_valid)
+}
+
+
+ds.build_glm1_formulas <- function(
+    meta_formula = NULL, # @todo: enable specification through formula 
+    variable_vectors = NULL, # @todo: meta_formula can reference named vectors
+    
+    dvs = c(),           # dependent variables each run independently
+    ivs = c(),           # independent variables each run independently
+    mods = c(""),        # moderators each run independently, default no mod
+    covs = c(),          # covariates included all together or none
+    cov_groups = list() # customize cov groups instead of default all or none
+    
+){
+    # Build vector of model formulas that meet the ds.glm1 spec
+    # for each unique combination of dv x iv x mod x cov_group.
+    # If covs supplied, cov_group gets one entry for no covs (unadjusted) 
+    # and one for all covs supplied in covs argument (adjusted).
+    
+    # @todo: to support meta_formula, make a new function to turn
+    # meta formulas into variable vectors and call on self
+    if( ! is.null(meta_formula) ){
+        util.warn("meta_formulas are not yet supported :-(")
+        # variable_list <- ds.meta_formula_to_variable_list(meta_formula)
+        # do.call(ds.build_glm1_formulas, variable_list) %>%
+        #   return()
+    } 
+    
+    if( length(dvs) == 0 | length(ivs) == 0 ){
+        util.warn("dvs and ivs must both be present.")
+    }
+    
+    if( length(covs) > 0 ){
+        # if covs is supplied, use it to define adjusted and unadjusted
+        # covariate groups with all and none of the covs, respectively
+        cov_groups <- list()
+        cov_groups[["Unadusted"]] <- ""
+        cov_groups[["Adjusted"]] <- covs
+    } else if( length(cov_groups) == 0){
+        cov_groups <- list( Unadjusted = "" )
+    }
+    
```

Comment:
It's nice to have the "adjusted" and "unadjusted" info here (there's a typo btw, "unadjusted" is missing a "j"), but it gets dropped down the line. By the time you run `ds.model_summary_df`, you get a data.frame with all the formulas you built here, which could be a lot if you have a lot of ivs and dvs. Sometimes you might not care about the unadjusted models and want to drop them, or you might want to be able to sort them to the top or something like that. So it could be useful to propogate this information forward so that the final summary df has an "adjusted" column.